consensus/tendermint/messages.go

```diff
@@ -0,0 +1,126 @@
+package tendermint
+
+// Todo: Signature over the messages needs to be handled somewhere. There are 2 options:
+//	1. Add the signature to each message and extend the Validator Set interface to include VerifyMessageSignature
+//	method.
+//	2. The P2P layer signs the message before gossiping to other validators and verifies the signature before passing
+//	the message to the consensus engine.
+//  The benefit of P2P layer handling the verification of the signature is the that the consensus layer can assume
+//  the message is from a validator in the validator set. However, this means that the P2P layer would need to be aware
+//  of the validator set and would need access to the blockchain which may not be a good idea.
+
+type Message[V Hashable[H], H Hash, A Addr] interface {
+	Proposal[V, H, A] | Prevote[H, A] | Precommit[H, A]
+}
+
+type Proposal[V Hashable[H], H Hash, A Addr] struct {
+	Height     uint
+	Round      uint
+	ValidRound *uint
+	Value      *V
+
+	Sender A
+}
+
+type Prevote[H Hash, A Addr] struct {
+	Vote[H, A]
+}
+
+type Precommit[H Hash, A Addr] struct {
+	Vote[H, A]
+}
+
+type Vote[H Hash, A Addr] struct {
+	Height uint
+	Round  uint
+	ID     *H
+
+	Sender A
+}
+
+// messages keep tracks of all the proposals, prevotes, precommits by creating a map structure as follows:
+// height->round->address->[]Message
+
+// Todo: would the following representation of message be better:
+//
+//	  height -> round -> address -> ID -> Message
+//	How would we keep track of nil votes? In golan map key cannot be nil.
+//	It is not easy to calculate a zero value when dealing with generics.
+type messages[V Hashable[H], H Hash, A Addr] struct {
+	proposals  map[uint]map[uint]map[A][]Proposal[V, H, A]
+	prevotes   map[uint]map[uint]map[A][]Prevote[H, A]
+	precommits map[uint]map[uint]map[A][]Precommit[H, A]
+}
+
+func newMessages[V Hashable[H], H Hash, A Addr]() messages[V, H, A] {
+	return messages[V, H, A]{
+		proposals:  make(map[uint]map[uint]map[A][]Proposal[V, H, A]),
+		prevotes:   make(map[uint]map[uint]map[A][]Prevote[H, A]),
+		precommits: make(map[uint]map[uint]map[A][]Precommit[H, A]),
+	}
+}
+
+// Todo: ensure duplicated messages are ignored.
+func (m *messages[V, H, A]) addProposal(p Proposal[V, H, A]) {
+	if _, ok := m.proposals[p.Height]; !ok {
+		m.proposals[p.Height] = make(map[uint]map[A][]Proposal[V, H, A])
+	}
+
+	if _, ok := m.proposals[p.Height][p.Round]; !ok {
+		m.proposals[p.Height][p.Round] = make(map[A][]Proposal[V, H, A])
+	}
+
+	sendersProposals, ok := m.proposals[p.Height][p.Round][p.Sender]
+	if !ok {
+		sendersProposals = []Proposal[V, H, A]{}
+	}
+
+	m.proposals[p.Height][p.Round][p.Sender] = append(sendersProposals, p)
```

Comment:
I guess we don't need to worry about this array being spammed by a malicious node? Does slashing depend on the number of equivocating messages / cap out at a certain point?

Reply:
We need to limit how many equivocated messages are stored as this is an attack vector. However, since the slashing conditions are unclear, I have not yet put a limit on them.

I think it is a design decision whether to slash more if a higher number of equivocated messages have been received. Sending even one should have a large consequence as slashing is the only incentive present to try to limit the number of faulty nodes to less than F.